craft_parts/parts.py

```diff
@@ -16,15 +16,95 @@
 
 """Definitions and helpers to handle parts."""
 
-import copy
 from pathlib import Path
 from typing import Any, Dict, List, Optional, Sequence, Set
 
+from pydantic import BaseModel, Field, ValidationError
+
 from craft_parts import errors
 from craft_parts.dirs import ProjectDirs
 from craft_parts.steps import Step
 
 
+class PartSpec(BaseModel):
+    """The part specification data."""
+
+    plugin: Optional[str]
+    source: Optional[str]
+    source_checksum: Optional[str] = ""
+    source_branch: Optional[str] = ""
+    source_commit: Optional[str] = ""
+    source_depth: Optional[int] = 0
+    source_subdir: Optional[str] = ""
+    source_tag: Optional[str] = ""
+    source_type: Optional[str] = ""
+    disable_parallel: Optional[bool] = False
+    after: Optional[List[str]] = []
```

Comment:
They make it work, but a default that appears mutable? yikes :D

Reply:
Given, `after: Optional[List[str]] = []`, I'd assume after can be None, yet the default is an empty list.

I'd go with the empty list default, and drop the Optional wherever the meaning is effectively the same.

This would avoid unnecessary handling like below:
```
 if not self.spec.after:
        return []
 return self.spec.after
```

Reply:
Pydantic handles list defaults in a sane way: https://stackoverflow.com/questions/63793662/how-to-give-a-pydantic-list-field-a-default-value (good). Unfortunately `Optional` becomes mandatory when unmarshaling with pydantic, otherwise it errors whenever the property is not listed in the input yaml (bad).